lint/base_linter/node_linter.py

```diff
@@ -146,21 +147,67 @@ def find_local_executable(self, start_dir, npm_name):
                     raise linter.PermanentError()
 
-                # A 'package.json' not yet installed?
-                if not os.path.exists(os.path.join(path, 'node_modules', '.bin')):
-                    is_dep = bool(manifest.get('dependencies', {}).get(npm_name))
-                    is_dev_dep = bool(manifest.get('devDependencies', {}).get(npm_name))
-                    if is_dep or is_dev_dep:
-                        logger.warning(
-                            "Skipping '{}' for now which is listed as a {} "
-                            "in {} but not installed.  Forgot to 'npm install'?"
-                            .format(
-                                npm_name,
-                                'dependency' if is_dep else 'devDependency',
-                                manifest_file
+                is_dep = bool(manifest.get('dependencies', {}).get(npm_name))
+                is_dev_dep = bool(manifest.get('devDependencies', {}).get(npm_name))
+                if is_dep or is_dev_dep:
+                    self.context['project_root'] = path
+
+                    # Perhaps this is a Yarn PnP project?
+                    yarn_path = shutil.which('yarn')
+                    yarn_lock_exists = os.path.exists(os.path.join(path, 'yarn.lock'))
+                    package_lock_exists = os.path.exists(os.path.join(path, 'package-lock.json'))
+                    pnp_js_exists = os.path.exists(os.path.join(path, '.pnp.js'))
+                    pnp_in_manifest = bool(manifest.get('installConfig', {}).get('pnp'))
+
+                    if yarn_path is None:
+                        if yarn_lock_exists and (pnp_in_manifest or pnp_js_exists):
+                            logger.warning(
+                                "This seems like a Yarn PnP project. However, finding "
+                                "a Yarn executable failed. Make sure to install Yarn first."
                             )
+                            self.notify_failure()
+                            raise linter.PermanentError()
+                    elif yarn_lock_exists and pnp_js_exists:
+                        return [yarn_path, 'run', '--silent', npm_name]
+
+                    next_path = os.path.dirname(path)
+
+                    for path in paths_upwards(next_path):
+                        executable = shutil.which(npm_name, path=os.path.join(path, 'node_modules', '.bin'))
+                        if executable:
+                            return executable
+
+                    logger.warning(
+                        "Skipping '{}' for now which is listed as a {} "
+                        "in {} but not installed.  Forgot to '{} install'?"
+                        .format(
+                            npm_name,
+                            'dependency' if is_dep else 'devDependency',
+                            manifest_file,
+                            'yarn' if yarn_path and yarn_lock_exists and not package_lock_exists else 'npm'
```

Comment:
Is having a yarn.lock and a package.json lock a real thing? What does that actually mean: I don't care use whatever you want? Or is that a config error on projects maintainer side?

Reply:
It *is* strongly discouraged to mix package managers like that, but still a user might checkout a repo that has yarn.lock and install the project using npm. I know yarn provides a warning for the mirror case but since npm is more widely used 🤷‍♀

Reply:
Then we don't have `yarn_path` defined here and we just don't test for it, just * yarn.lock* but not *packagejson.lock* should be ok?